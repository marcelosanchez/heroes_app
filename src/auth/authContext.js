import { createContext } from "react";

export const AuthContext = createContext();  // exporto el contexto de autenticacion
