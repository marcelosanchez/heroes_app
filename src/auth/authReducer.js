import { types } from "../types/types";

// const state = {
//     name: 'Marco',
//     logged: true,
// }

// const loginAction = {
//     type: types.login,
//     payload: {
//         name: 'Marco',
//         email: 'marco@mail.com,'
//     }
// }

export const authReducer = ( state = {}, action ) => {  // funcion reductora
    switch ( action.type ) {
        case types.login:  // si el tipo es login
            return {
                ...action.payload,  // retorno el payload con el nombre
                logged: true,
            }
        case types.logout:  // si el tipo es logout
            return {
                logged: false,
            }
        default:
            return state;  // si no hay ninguna accion, devuelvo el estado
    }
};