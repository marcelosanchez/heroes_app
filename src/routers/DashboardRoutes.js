import React from 'react'
import { Route, Routes } from 'react-router-dom'
import { Navbar } from '../components/ui/Navbar'

import { DcScreen } from '../components/dc/DcScreen';
import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { SearchScreen } from '../components/search/SearchScreen';
import { HeroScreen } from '../components/hero/HeroScreen';

export const DashboardRoutes = () => {
    return (
        <>
            <Navbar />
            
            <div className='container pt-3'>
                <Routes>
                    <Route path="marvel" element={<MarvelScreen />} />
                    <Route path="dc" element={<DcScreen />} />

                    <Route path="search" element={<SearchScreen />} />
                    <Route path="hero/:heroId" element={<HeroScreen/>} />

                    <Route path="/" element={<MarvelScreen />} />
                </Routes>
            </div>
        </>
    )
}
