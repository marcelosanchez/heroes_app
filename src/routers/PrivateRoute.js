import React, { useContext, useEffect } from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import { AuthContext } from '../auth/authContext';


export const PrivateRoute = ( { children } ) => {

    const { user } = useContext(AuthContext);

    const location = useLocation();  // se obtiene la ruta actual
    const lastPath = location.pathname + location.search;  // se obtiene la ruta actual con query string

    useEffect(() => {
        localStorage.setItem('lastpath', lastPath);  // se guarda la ultima ruta visitada
    }, [ location ]);  // se ejecuta solo cuando la ruta cambia

    return (
            user.logged 
            ? children 
            : <Navigate to='/login' />
    )
}
