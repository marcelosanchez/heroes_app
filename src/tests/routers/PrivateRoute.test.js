import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { PrivateRoute } from '../../routers/PrivateRoute';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    Navigate: () => <span>Saliendo de aqui</span>  // El componente Navigate es una funci√≥n que se utiliza para redireccionar
}));

describe('Pruebas en <PrivateRoute />', () => {

    Storage.prototype.setItem = jest.fn();  // permite guardar el valor en el localStorage

    test('Debe mostrar el componente si esta autenticado y guardar en el localStorage', () => {
        const contextValue = {
            user: {
                logged: true,
                name: 'Marcelo',
            }
        };

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/']}>
                    <PrivateRoute>
                        <h1>Private Component</h1>
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.text().trim() ).toBe('Private Component');
        expect( localStorage.setItem ).toHaveBeenCalledWith('lastpath', '/');  // se guarda la ultima ruta visitada
    })

    test('Debe bloquear el componente si no esta autenticado', () => {
        const contextValue = {
            user: {
                logged: false,
            }
        };

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/']}>
                    <PrivateRoute>
                        <h1>Private Component</h1>
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.text().trim() ).toBe('Saliendo de aqui');
    })
})