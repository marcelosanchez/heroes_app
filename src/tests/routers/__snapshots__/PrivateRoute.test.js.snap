// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <PrivateRoute /> Debe bloquear el componente si no esta autenticado 1`] = `
<span>
  Saliendo de aqui
</span>
`;

exports[`Pruebas en <PrivateRoute /> Debe mostrar el componente si esta autenticado y guardar en el localStorage 1`] = `
<h1>
  Private Component
</h1>
`;
